src/Authentication/ApiKey/Server.php

```diff
@@ -108,17 +109,14 @@ public function authenticate( $user_id ) {
 	 *
 	 * @since 0.3.0
 	 *
-	 * @param WP_Error $error Error object.
+	 * @param HttpException $e HTTP exception.
 	 */
-	protected function handle_error( WP_Error $error ) {
-		$error_data = $error->get_error_data();
-
-		if ( ! empty( $error_data['status'] ) && HTTP::UNAUTHORIZED === $error_data['status'] ) {
+	public function handle_error( HttpException $e ): WP_Error {
+		if ( HTTP::UNAUTHORIZED === $e->getStatusCode() ) {
 			header( 'WWW-Authenticate: Basic realm="SatisPress"' );
 		}
 
-		$status_code = empty( $error_data['status'] ) ? HTTP::INTERNAL_SERVER_ERROR : $error_data['status'];
-		wp_die( wp_kses_data( $error->get_error_message() ), absint( $status_code ) );
+		wp_die( wp_kses_data( $e->getMessage() ), absint( $e->getStatusCode() ) );
```

Comment:
Is there a way to let another class handle this sanitization and dieing, so that there is no need to link this class to WP code?
(There's an `is_wp_error()` further down as well).

Reply:
I've been thinking on this and can't really come up with a solution that isn't convoluted. The authentication handling was designed to be compatible with the REST API in case REST routes are registered in the future. The current `wp_die()` approach isn't really compatible with that, but neither is throwing an exception since [authentication errors are retrieved](https://github.com/WordPress/WordPress/blob/6b6201ed8937f56ff2d2e65c387db834e76b78f5/wp-includes/rest-api/class-wp-rest-server.php#L135) through the `rest_authentication_errors` filter, which expects `WP_Error`, `null`, or `true`. It's also not possible to send headers through `WP_Error`.

Hmm, maybe the headers can be added when the `HttpException` is created in the `authenticate()` method. Then the `handle_error()` method can be dropped. `Authentication::get_authentication_errors()` can then convert the exception into an error object for REST requests, or rethrow the exception for the request handler to convert it into a response.

Reply:
@GaryJones Let me know what you think of those changes. I'm not sure if the authentication exceptions are doing too much or not. This does simplify the logic in the servers quite a bit, though.